qiskit/synthesis/clifford/clifford_decompose_layers.py

```diff
@@ -0,0 +1,408 @@
+    layers is a list of QuantumCircuits
+    """
+
+    def __init__(self, num_qubits: int):
+        """Stores layered decomposition of the QuantumCircuit
+
+        Args:
+            num_qubits: number of qubits of the QuantumCircuit.
+        """
+        self.num_qubits = num_qubits
+        self.layers = []
+
+    def append_layer(self, qc, qc_name=None):
+        """Appends a new layer.
+        qc is a quantum circuit
+        qc_name is the optional label to assign to qc
+        """
+        if not isinstance(qc, QuantumCircuit):
+            raise QiskitError("qc is not a QuantumCircuit.")
+        if qc.num_qubits != self.num_qubits:
+            raise QiskitError("num_qubits is not the same as qc.num_qubits.")
+
+        if qc_name is not None:
+            qc.name = qc_name
+        self.layers.append(qc)
+
+    def create_circuit(self):
+        """Returns circuit obtained by appending all layers"""
+        circ = QuantumCircuit(self.num_qubits)
+        for _, qc in enumerate(list(self.layers)):
+            circ.append(qc, list(range(self.num_qubits)))
+        return circ
+
+    def draw(self):
+        """Print the layered circuit."""
+        print(self.create_circuit())
+
+    def draw_detailed(self):
+        """Print the circuit layer-by-layer"""
+        print()
+        for i, qc in enumerate(list(self.layers)):
+            print(f"Printing layer {i}")
+            print(qc)
+            print("")
+
+
+def synth_clifford_layers(
+    cliff,
+    cx_synth_func,
+    cz_synth_func,
+    cx_cz_synth_func=None,
+    reverse_cliff=False,
+    validate=False,
+):
+    """Synthesis of a Clifford into layers, it provides a similar decomposition to the synthesis
+    described in Lemma 8 of [1].
+
+    For example, a 5-qubit Clifford is decomposed into the following layers:
+
+         ┌─────┐┌──────┐┌─────┐┌─────┐┌─────┐┌──────┐┌─────┐┌────────┐
+    q_0: ┤0    ├┤0     ├┤0    ├┤0    ├┤0    ├┤0     ├┤0    ├┤0       ├
+         │     ││      ││     ││     ││     ││      ││     ││        │
+    q_1: ┤1    ├┤1     ├┤1    ├┤1    ├┤1    ├┤1     ├┤1    ├┤1       ├
+         │  S2 ││  CZ2 ││  CX ││  H2 ││  S1 ││  CZ1 ││  H1 ││  Pauli │
+    q_2: ┤2    ├┤2     ├┤2    ├┤2    ├┤2    ├┤2     ├┤2    ├┤2       ├
+         │     ││      ││     ││     ││     ││      ││     ││        │
+    q_3: ┤3    ├┤3     ├┤3    ├┤3    ├┤3    ├┤3     ├┤3    ├┤3       ├
+         └─────┘└──────┘└─────┘└─────┘└─────┘└──────┘└─────┘└────────┘
+
+    Args:
+        cliff (Clifford): a clifford operator.
+        cx_synth_func (Callable): a function to decompose the CX sub-circuit.
+        cz_synth_func (Callable): a function to decompose the CZ sub-circuit.
+        cx_cz_synth_func (Callable): optional, a function to decompose the sub-circuits CZ and CX .
+        reverse_cliff (Boolean): if True, reverse the order of the qubits of the Clifford.
+        validate (Boolean): if True, validates the synthesis process.
```

Comment:
I am actually still confused when we need to reverse the order of the qubits (for the option `reverse_cliff`). Can we remove this option, and instead let the calling application do this reversal when required?

Reply:
One more minor comment. Can you also specify the default values for `cx_synth_func` and `cz_synth_func`, i. e. 
```
def synth_clifford_layers(
    cliff,
    cx_synth_func=_default_cx_synth_func,
    cz_synth_func=_default_cz_synth_func,
    cx_cz_synth_func=None,
    reverse_cliff=False,
    validate=False,
):
```